pymc/logprob/censoring.py

```diff
@@ -238,3 +260,282 @@ def round_logprob(op, values, base_rv, **kwargs):
+        for i in range(2):
+            switch_dict = {
+                "lower": adjusted_intervals[i][0],
+                "upper": adjusted_intervals[i][1],
+                "encoding": components[i],
+            }
+            encoding_list.append(switch_dict)
+
+        return encoding_list
+
+    else:
+        for i in range(2):
+            if i in switch_comp_idx:
+                # Recurse through the switch component(es)
+                encoding_list = flat_switch_helper(
+                    components[i].owner, valued_rvs, encoding_list, adjusted_intervals[i], base_rv
+                )
+
+            else:
+                switch_dict = {
+                    "lower": adjusted_intervals[i][0],
+                    "upper": adjusted_intervals[i][1],
+                    "encoding": components[i],
+                }
+                encoding_list.append(switch_dict)
+
+    return encoding_list
+
+
+@node_rewriter(tracks=[switch])
+def find_measurable_flat_switch_encoding(fgraph: FunctionGraph, node: Node):
+    rv_map_feature: Optional[PreserveRVMappings] = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return None  # pragma: no cover
+
+    valued_rvs = rv_map_feature.rv_values.keys()
+    switch_cond = node.inputs[0]
+
+    encoding_list = []
+    initial_interval = (-np.inf, np.inf)
+
+    # fetch base_var as the only measurable input to the logical op in switch condition
+    measurable_switch_inp = [
+        component
+        for component in switch_cond.owner.inputs
+        if check_potential_measurability([component], valued_rvs)
+    ]
+
+    if len(measurable_switch_inp) != 1:
+        return None
+
+    base_rv = measurable_switch_inp[0]
+
+    # We do not allow discrete RVs yet
+    if base_rv.dtype.startswith("int"):
+        return None
+
+    # Since we verify the source of measurability to be the same for switch conditions
+    # and all measurable components, denying broadcastability of the base_var is enough
+    if base_rv.type.broadcastable != node.outputs[0].type.broadcastable:
+        return None
+
+    encoding_list = flat_switch_helper(node, valued_rvs, encoding_list, initial_interval, base_rv)
+    if encoding_list is None:
+        return None
+
+    encodings, intervals = [], []
+    rv_idx = ()
+
+    # TODO: Some alternative cleaner way to do this
+    for idx, item in enumerate(encoding_list):
+        encoding = item["encoding"]
+        # indices of intervals having base_rv as their "encoding"
+        if encoding == base_rv:
+            rv_idx += (idx,)
+
+        encodings.append(encoding)
+        intervals.extend((item["lower"], item["upper"]))
+
+    flat_switch_op = FlatSwitches(out_dtype=node.outputs[0].dtype, rv_idx=rv_idx)
+
+    new_outs = flat_switch_op.make_node(base_rv, *intervals, *encodings).default_output()
+    return [new_outs]
+
+
+@_logprob.register(FlatSwitches)
+def flat_switches_logprob(op, values, base_rv, *inputs, **kwargs):
```

Comment:
```suggestion
def nested_switch_encoding_logprob(op, values, base_rv, *inputs, **kwargs):
```

Reply:
Also add some comment in the docstrings about the kind of logp graphs we are generating?